bus-mapping/src/evm/opcodes/coinbase.rs

```diff
@@ -0,0 +1,91 @@
+use super::Opcode;
+use crate::circuit_input_builder::CircuitInputStateRef;
+use crate::eth_types::GethExecStep;
+use crate::{operation::RW, Error};
+
+/// Placeholder structure used to implement [`Opcode`] trait over it
+/// corresponding to the [`OpcodeId::PC`](crate::evm::OpcodeId::PC) `OpcodeId`.
+#[derive(Debug, Copy, Clone)]
+pub(crate) struct Coinbase;
+
+impl Opcode for Coinbase {
+    fn gen_associated_ops(
+        state: &mut CircuitInputStateRef,
+        steps: &[GethExecStep],
+    ) -> Result<(), Error> {
+        let step = &steps[0];
+        // Get value result from next step and do stack write
+        let value = steps[1].stack.last()?;
+        state.push_stack_op(
+            RW::WRITE,
+            step.stack.last_filled().map(|a| a - 1),
+            value,
+        );
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod coinbase_tests {
+    use super::*;
+    use crate::{
+        bytecode,
+        circuit_input_builder::{ExecStep, TransactionContext},
+        eth_types::ToWord,
+        evm::StackAddress,
+        mock,
+    };
+    use pretty_assertions::assert_eq;
+
+    #[test]
+    fn coinbase_opcode_impl() -> Result<(), Error> {
+        let code = bytecode! {
+            #[start]
+            COINBASE
+            STOP
+        };
+
+        // Get the execution steps from the external tracer
+        let block =
+            mock::BlockData::new_single_tx_trace_code_at_start(&code).unwrap();
+
+        let mut builder = block.new_circuit_input_builder();
+        builder.handle_tx(&block.eth_tx, &block.geth_trace).unwrap();
+
+        let mut test_builder = block.new_circuit_input_builder();
+        let mut tx = test_builder.new_tx(&block.eth_tx).unwrap();
+        let mut tx_ctx = TransactionContext::new(&block.eth_tx);
+
+        // Generate step corresponding to COINBASE
+        let mut step = ExecStep::new(
+            &block.geth_trace.struct_logs[0],
+            0,
+            test_builder.block_ctx.rwc,
+            0,
+        );
+        let mut state_ref =
+            test_builder.state_ref(&mut tx, &mut tx_ctx, &mut step);
+
+        // Add the last Stack write
+        state_ref.push_stack_op(
+            RW::WRITE,
+            StackAddress::from(1024 - 1),
+            block.ctants.coinbase.to_word(),
```

Comment:
I want to refactor  to  use `BlockConstants`'s coinbase field  here and remove `ChainConstants` coinbase field, not sure why exists in chain constant of bussmapping before, want to hear from you guys first @ChihChengLiang  @han0110 @CPerezz

Reply:
Maybe @ed255 can provide some insight on the bus-mapping related stuff.
From my perspective, the coinbase will always be the same. And therefore is only passed once via `ChainConstants` instead of passing it on each tx/block.

Reply:
coinbase is only block's field not related to tx， each block can have different `coinbase` address,  chainid is global which make sense exists in `ChainConstants`

Reply:
I originally put the `coinbase` in `ChainConstants`.  I had the wrong idea of what the `coinbase` was, so it was a mistake.  `coinbase` should be removed from `ChainConstants`.

The `coinbase` that you need here I think should be `block.eth_block.author`. See documentation here https://docs.rs/ethers/latest/ethers/core/types/struct.Block.html#structfield.author
Block `miner`, `author`, `coinbase` are all the same thing.

The struct `BlockConstants` defined in `external_tracer` has a name which is a bit misleading.  That struct is really a configuration parameter to setup an environment for the external tracer.

Reply:
Ohh my bad... 

I understood coinbase on a different way...

Reply:
thanks @ed255 , I will try remove `coinbase` in `ChainConstants` at this PR!

Reply:
fixed in 21ad52f